File: AI-Reply/Shared/AI/ReplyDraftNormalizer.swift
import Foundation

/// Final lightweight cleanup before a draft is inserted into the host field.
/// URL substrings are copied through untouched.
struct ReplyDraftNormalizer {

    private static let urlRegex = try! NSRegularExpression(
        pattern: #"(?i)\b(?:https?://|www\.)\S+"#
    )
    private static let repeatedHorizontalSpaceRegex = try! NSRegularExpression(
        pattern: #"[^\S\r\n]{2,}"#
    )
    private static let spaceBeforePunctuationRegex = try! NSRegularExpression(
        pattern: #"[^\S\r\n]+([,.!?;:])"#
    )
    private static let excessiveBlankLineRegex = try! NSRegularExpression(
        pattern: #"\n[ \t]*\n(?:[ \t]*\n)+"#
    )

    func normalize(_ draft: String) -> String {
        let trimmed = draft.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return "" }

        var output = ""
        let nsText = trimmed as NSString
        var cursor = 0
        let fullRange = NSRange(location: 0, length: nsText.length)

        for match in Self.urlRegex.matches(in: trimmed, range: fullRange) {
            if match.range.location > cursor {
                let plainRange = NSRange(location: cursor, length: match.range.location - cursor)
                output += Self.normalizePlainSegment(nsText.substring(with: plainRange))
            }

            output += nsText.substring(with: match.range)
            cursor = match.range.location + match.range.length
        }

        if cursor < nsText.length {
            let plainRange = NSRange(location: cursor, length: nsText.length - cursor)
            output += Self.normalizePlainSegment(nsText.substring(with: plainRange))
        }

        return Self.excessiveBlankLineRegex.stringByReplacingMatches(
            in: output,
            range: NSRange(location: 0, length: (output as NSString).length),
            withTemplate: "\n\n"
        )
        .trimmingCharacters(in: .whitespacesAndNewlines)
    }

    private static func normalizePlainSegment(_ segment: String) -> String {
        var value = repeatedHorizontalSpaceRegex.stringByReplacingMatches(
            in: segment,
            range: NSRange(location: 0, length: (segment as NSString).length),
            withTemplate: " "
        )

        value = spaceBeforePunctuationRegex.stringByReplacingMatches(
            in: value,
            range: NSRange(location: 0, length: (value as NSString).length),
            withTemplate: "$1"
        )

        return value
    }
}

File: AI-Reply-Android/app/src/main/java/kz/yerek/aireply/ai/ReplyDraftNormalizer.kt
package kz.yerek.aireply.ai

/**
 * Final lightweight cleanup before a draft is inserted into the host field.
 *
 * URL substrings are copied through untouched, which is the whole reason this
 * is not a single chain of replaces: collapsing double spaces inside a URL, or
 * removing the space before the punctuation in one, produces a link that does
 * not work.
 */
class ReplyDraftNormalizer {

    fun normalize(draft: String): String {
        val trimmed = draft.trim()
        if (trimmed.isEmpty()) return ""

        val output = StringBuilder()
        var cursor = 0

        for (match in URL.findAll(trimmed)) {
            if (match.range.first > cursor) {
                output.append(normalizePlain(trimmed.substring(cursor, match.range.first)))
            }
            output.append(match.value)
            cursor = match.range.last + 1
        }

        if (cursor < trimmed.length) {
            output.append(normalizePlain(trimmed.substring(cursor)))
        }

        return EXCESSIVE_BLANK_LINES.replace(output.toString(), "\n\n").trim()
    }

    private fun normalizePlain(segment: String): String {
        var value = REPEATED_SPACE.replace(segment, " ")
        value = SPACE_BEFORE_PUNCTUATION.replace(value) { it.groupValues[1] }
        return value
    }

    private companion object {
        val URL = Regex("""\b(?:https?://|www\.)\S+""", RegexOption.IGNORE_CASE)
        val REPEATED_SPACE = Regex("""[^\S\r\n]{2,}""")
        val SPACE_BEFORE_PUNCTUATION = Regex("""[^\S\r\n]+([,.!?;:])""")
        val EXCESSIVE_BLANK_LINES = Regex("""\n[ \t]*\n(?:[ \t]*\n)+""")
    }
}
